prelude/slang-cpp-types.h

```diff
@@ -228,136 +228,132 @@ struct SamplerComparisonState
-    virtual void SampleLevel(SamplerState samplerState, const float3& loc, float level, void* out) = 0;
+    T Sample(SamplerState samplerState, const float3& loc) const { T out; texture->Sample(samplerState, &loc.x, &out); return out; }
+    T SampleLevel(SamplerState samplerState, const float3& loc, float level) { T out; texture->SampleLevel(samplerState, &loc.x, level, &out); return out; }
+    
+    ITexture* texture;              
 };
 
 template <typename T>
-struct TextureCube
+struct Texture1DArray
 {
-    T Sample(SamplerState samplerState, const float3& v) const { T out; texture->Sample(samplerState, v, &out); return out; }
-    T SampleLevel(SamplerState samplerState, const float3& v, float level) { T out; texture->SampleLevel(samplerState, v, level, &out); return out; }
+    T Load(const int3& loc) const { T out; texture->Load(&loc.x, &out); return out; }
+    T Sample(SamplerState samplerState, const float2& loc) const { T out; texture->Sample(samplerState, &loc.x, &out); return out; }
+    T SampleLevel(SamplerState samplerState, const float2& loc, float level) { T out; texture->SampleLevel(samplerState, &loc.x, level, &out); return out; }
     
-    ITextureCube* texture;              
+    ITexture* texture;              
 };
 
-struct ITexture1DArray
+template <typename T>
+struct Texture2DArray
 {
-    virtual void Load(const int3& v, void* out) = 0;
-    virtual void Sample(SamplerState samplerState, const float2& loc, void* out) = 0;
-    virtual void SampleLevel(SamplerState samplerState, const float2& loc, float level, void* out) = 0;
+    T Load(const int4& loc) const { T out; texture->Load(&loc.x, &out); return out; }
+    T Sample(SamplerState samplerState, const float3& loc) const { T out; texture->Sample(samplerState, &loc.x, &out); return out; }
+    T SampleLevel(SamplerState samplerState, const float3& loc, float level) { T out; texture->SampleLevel(samplerState, &loc.x, level, &out); return out; }
+    
+    ITexture* texture;              
 };
 
 template <typename T>
-struct Texture1DArray
+struct TextureCubeArray
 {
-    T Load(const int3& v) const { T out; texture->Load(v, &out); return out; }
-    T Sample(SamplerState samplerState, const float2& v) const { T out; texture->Sample(samplerState, v, &out); return out; }
-    T SampleLevel(SamplerState samplerState, const float2& v, float level) { T out; texture->SampleLevel(samplerState, v, level, &out); return out; }
+    T Sample(SamplerState samplerState, const float4& loc) const { T out; texture->Sample(samplerState, &loc.x, &out); return out; }
+    T SampleLevel(SamplerState samplerState, const float4& loc, float level) { T out; texture->SampleLevel(samplerState, &loc.x, level, &out); return out; }
     
-    ITexture1DArray* texture;              
+    ITexture* texture;              
 };
 
-struct ITexture2DArray
+/* !!!!!!!!!!!!!!!!!!!!!!!!!!! RWTexture !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! */
+
+struct IRWTexture
```

Comment:
It is correct/accurate that `RWTexture2D` would not have the `Sample` operations and similar stuff from `Texture2D`? Otherwise, I would kind of expect `IRWTexture` to inherit from `ITexture`

Reply:
Yes the RWTexture seems like it should derive from 'read only' Texture, but it is sort of something different, and doesn't support sampleState. It's the same under CUDA with surfaces.